rom/dev/tests/rom_integration_tests/test_warm_reset.rs

```diff
@@ -0,0 +1,236 @@
+    let security_state = *SecurityState::default()
+        .set_debug_locked(true)
+        .set_device_lifecycle(DeviceLifecycle::Production);
+
+    let rom = caliptra_builder::build_firmware_rom(&ROM_WITH_UART).unwrap();
+    let image = caliptra_builder::build_and_sign_image(
+        &FMC_WITH_UART,
+        &APP_WITH_UART,
+        ImageOptions {
+            fmc_min_svn: 5,
+            fmc_svn: 9,
+            ..Default::default()
+        },
+    )
+    .unwrap();
+    let vendor_pk_hash =
+        bytes_to_be_words_48(&sha384(image.manifest.preamble.vendor_pub_keys.as_bytes()));
+    let owner_pk_hash =
+        bytes_to_be_words_48(&sha384(image.manifest.preamble.owner_pub_keys.as_bytes()));
+
+    let mut hw = caliptra_hw_model::new(BootParams {
+        init_params: InitParams {
+            rom: &rom,
+            security_state,
+            ..Default::default()
+        },
+        fuses: Fuses {
+            key_manifest_pk_hash: vendor_pk_hash,
+            owner_pk_hash,
+            fmc_key_manifest_svn: 0b1111111,
+            ..Default::default()
+        },
+        fw_image: Some(&image.to_bytes().unwrap()),
+        ..Default::default()
+    })
+    .unwrap();
+
+    // Wait for boot
+    while !hw.soc_ifc().cptra_flow_status().read().ready_for_runtime() {
+        hw.step();
+    }
+
+    // Perform warm reset
+    hw.warm_reset_flow(&Fuses {
+        key_manifest_pk_hash: vendor_pk_hash,
+        owner_pk_hash,
+        fmc_key_manifest_svn: 0b1111111,
+        ..Default::default()
+    });
+
+    // Wait for boot
+    while !hw.soc_ifc().cptra_flow_status().read().ready_for_runtime() {
+        hw.step();
+    }
+}
+
+#[test]
+fn test_warm_reset_during_cold_boot_before_image_validation() {
+    let fuses = Fuses {
+        life_cycle: DeviceLifecycle::Production,
+        ..Default::default()
+    };
+
+    let (mut hw, _image_bundle) =
+        helpers::build_hw_model_and_image_bundle(fuses, ImageOptions::default());
+
+    // Step till Cold boot starts
+    hw.step_until_boot_status(IDevIdDecryptUdsComplete.into(), true);
+
+    // Perform a warm reset
+    hw.warm_reset_flow(&Fuses::default());
+
+    // Wait for error
+    while hw.soc_ifc().cptra_fw_error_fatal().read() == 0 {
+        hw.step();
+    }
+    assert_eq!(
+        hw.soc_ifc().cptra_fw_error_fatal().read(),
+        u32::from(CaliptraError::ROM_WARM_RESET_UNSUCCESSFUL_PREVIOUS_COLD_RESET)
+    );
+}
+
+#[test]
+fn test_warm_reset_during_cold_boot_during_image_validation() {
+    let fuses = Fuses {
+        life_cycle: DeviceLifecycle::Unprovisioned,
+        ..Default::default()
+    };
+
+    let (mut hw, image_bundle) =
+        helpers::build_hw_model_and_image_bundle(fuses, ImageOptions::default());
+
+    // Manually put the firmware in the mailbox because
+    // HwModel::upload_firmware returns only when the transaction is complete.
+    // This is too late for this test.
+    let buf: &[u8] = &image_bundle.to_bytes().unwrap();
```

Comment:
Would hw.start_mailbox_execute(FW_LOAD_CMD_OPCODE, ...) work?

Reply:
That places the command in the mailbox but doesn't wait for it to complete.l

Reply:
Updated.